feat: auction engine V4

nieldeckx opened the pull request.

philipdesmedt commented on clarity/contracts/arkadiko-auction-engine-v4-1.clar:
```diff
@@ -0,0 +1,295 @@
+;; @contract Auction Engine - Sells off vault collateral to raise USDA
+;; @version 1
```
NTH: update version

philipdesmedt commented on clarity/contracts/arkadiko-auction-engine-v4-1.clar:
```diff
@@ -0,0 +1,295 @@
+    (asserts! (is-eq (contract-of liquidation-pool) (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "liquidation-pool"))) (err ERR-NOT-AUTHORIZED))
+    (asserts! (is-eq (contract-of liquidation-rewards) (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "liquidation-rewards"))) (err ERR-NOT-AUTHORIZED))
+
+    (asserts! (or (is-eq (contract-of ft) token-address) (is-eq "xSTX" token-string)) (err ERR-TOKEN-TYPE-MISMATCH))
```
Q: why do you need to xSTX comparison? xSTX is also a SIP10 FT that has a `token-address`, no? In that case, we would expect the `ft` argument to be `.wrapped-stx-token`?

philipdesmedt commented on clarity/contracts/arkadiko-auction-engine-v4-1.clar:
```diff
@@ -0,0 +1,295 @@
+    (asserts! (is-eq (contract-of coll-type) (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "collateral-types"))) (err ERR-NOT-AUTHORIZED))
+    (asserts! (is-eq (contract-of oracle) (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "oracle"))) (err ERR-NOT-AUTHORIZED))
+    (asserts! (is-eq (contract-of reserve) (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "sip10-reserve"))) (err ERR-NOT-AUTHORIZED))
+    (asserts! (is-eq (get is-liquidated vault) false) (err ERR-AUCTION-NOT-ALLOWED))
```
NTH: prefer `(not (get is-liquidated vault))`

nieldeckx replied:
Agree.

philipdesmedt commented on clarity/contracts/arkadiko-auction-engine-v4-1.clar:
```diff
@@ -0,0 +1,295 @@
+    (asserts! (is-eq (contract-of vault-manager) (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "freddie"))) (err ERR-NOT-AUTHORIZED))
+    (asserts! (is-eq (contract-of coll-type) (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "collateral-types"))) (err ERR-NOT-AUTHORIZED))
+    (asserts! (is-eq (contract-of oracle) (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "oracle"))) (err ERR-NOT-AUTHORIZED))
+    (asserts! (is-eq (contract-of reserve) (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "sip10-reserve"))) (err ERR-NOT-AUTHORIZED))
```
REQ: can't this be `stx-reserve` as well? If a vault is not stacking, we auction off STX straight away which is stored by stx-reserve

philipdesmedt commented on clarity/contracts/arkadiko-auction-engine-v4-1.clar:
```diff
@@ -0,0 +1,295 @@
+;; @param reserve; reserve to get collateral tokens from
+;; @param liquidation-pool; pool to get USDA from
+;; @param liquidation-rewards; contract to deposit rewards
+(define-public (burn-usda
```
Q: should this be publicly callable?

nieldeckx replied:
This is useful when there was not enough USDA in the pool to fully liquidate the vault when starting the auction. So part of the liquidation can be done afterwards.

philipdesmedt commented on clarity/contracts/arkadiko-auction-engine-v4-1.clar:
```diff
@@ -0,0 +1,295 @@
+    (asserts! (is-eq (contract-of vault-manager) (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "freddie"))) (err ERR-NOT-AUTHORIZED))
+    (asserts! (is-eq (contract-of coll-type) (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "collateral-types"))) (err ERR-NOT-AUTHORIZED))
+    (asserts! (is-eq (contract-of oracle) (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "oracle"))) (err ERR-NOT-AUTHORIZED))
+    (asserts! (is-eq (contract-of reserve) (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "sip10-reserve"))) (err ERR-NOT-AUTHORIZED))
```
REQ: can't this be stx-reserve as well? If a vault is not stacking, we auction off STX straight away which is stored by stx-reserve

philipdesmedt commented on clarity/contracts/arkadiko-dao.clar:
```diff
@@ -416,11 +416,56 @@
     {
       address: 'ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM,
-      qualified-name: 'ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-liquidator-v2-1
+      qualified-name: 'ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-auction-engine-v4-1
```
Q: why has the liquidator become the auction engine?

Also I don't see the `liquidate` function called anywhere on `arkadiko-freddie-v1-1` anymore, which is necessary to tag the vault as liquidated and for the user to eventually withdraw leftover collateral after the auction

nieldeckx replied:
The auction-manager will:
1) start the auction and call `liquidate` on `arkadiko-freddie-v1-1`
2) burn USDA and get the some of the collateral to deposit to `arkadiko-liquidation-rewards-v1-1`
3) finish the auction by calling `finalize-liquidation` on `arkadiko-freddie-v1-1`

No need to have a separate liquidator imo. All liquidation logic is now part of `arkadiko-auction-engine-v4-4`

philipdesmedt reviewed (changes requested):
Did a first pass

nieldeckx commented on clarity/contracts/arkadiko-auction-engine-v4-1.clar:
```diff
@@ -0,0 +1,295 @@
+    )
+
+    ;; Try to finalize
+    (unwrap-panic (finalize-liquidation auction-id vault-manager coll-type))
```
My bad, changed.

nieldeckx commented on clarity/contracts/arkadiko-liquidation-rewards-diko-v1-1.clar:
```diff
@@ -0,0 +1,53 @@
+(define-constant blocks-per-epoch u2016) ;; 2 weeks
+
+;; Variables
+(define-data-var end-epoch-block uint (+ block-height blocks-per-epoch))
```
Yes, done.

nieldeckx commented on clarity/contracts/arkadiko-liquidation-rewards-diko-v1-1.clar:
```diff
@@ -0,0 +1,53 @@
+    (total-block-rewards (contract-call? .arkadiko-diko-guardian-v1-1 get-staking-rewards-per-block))
+
+    ;; 10% of staking rewards
+    (rewards-to-add (/ (* total-block-rewards blocks-per-epoch) u100000))
```
We could easily deploy a new contract `arkadiko-liquidation-rewards-diko-v2-1` to update the rate and/or epoch length. 
But will also add 2 new methods to easily change both variables.

nieldeckx replied:
Done

nieldeckx commented on clarity/contracts/arkadiko-dao.clar:
```diff
@@ -416,11 +416,56 @@
+    }
+  )
+
+  (map-set contracts
```
Done.

nieldeckx commented:
@philipdesmedt contracts ready for review. Web still draft.

philipdesmedt commented on clarity/contracts/arkadiko-liquidation-rewards-diko-v1-1.clar:
```diff
@@ -0,0 +1,106 @@
+)
+
+;; @desc add DIKO rewards to liquidation-rewards
+(define-public (add-rewards (liquidation-rewards <liquidation-rewards-trait>))
```
Q: we/someone has to call this once per epoch?

nieldeckx replied:
Correct. Can be done retroactively if rewards were not added during epoch.

philipdesmedt commented on clarity/contracts/arkadiko-liquidation-rewards-v1-1.clar:
```diff
@@ -0,0 +1,186 @@
+      (begin 
+        (if (get token-is-stx reward-info)
+          (try! (as-contract (stx-transfer? reward-amount (as-contract tx-sender) user)))
+          (try! (as-contract (contract-call? token transfer reward-amount (as-contract tx-sender) user none)))
```
REQ: I think this is a severe vulnerability, as follows

1. Anyone can add any SIP10 token, meaning that I could implement a malicious FT that transfers a bunch of STX in its transfer function.
2. I start staking in the pool and add 1 token of this as a reward (`add-reward` function)
3. I wait until the epoch is over
4. I run this function (`claim-rewards-of`)
5. Since I am eligible for the rewards, my malicious transfer function will run and drain all the tokens that are kept on this contract (could be xBTC, STX, ...)

philipdesmedt replied:
Solution is simple

Add a map with key principal `(contract-of token)` that keeps a whitelist and checks that in an assertion when adding and claiming rewards

nieldeckx replied:
Right, great find! Will apply the solution.

philipdesmedt commented on clarity/contracts/arkadiko-auction-engine-v4-1.clar:
```diff
@@ -0,0 +1,396 @@
+;; @param reserve; reserve to get collateral tokens from
+;; @param liquidation-pool; pool to get USDA from
+;; @param liquidation-rewards; contract to deposit rewards
+(define-public (burn-usda
```
REQ: this can be called on any auction, we should check that the auction is still open.

nieldeckx replied:
If the auction is already closed, there is no collateral to sell and `usda-to-use` will be 0.

philipdesmedt commented on clarity/contracts/arkadiko-auction-engine-v4-1.clar:
```diff
@@ -0,0 +1,396 @@
+  (let (
+    (auction (get-auction-by-id auction-id))
+    (all-collateral-sold (is-eq (get collateral-amount auction) (get total-collateral-sold auction)))
+    (debt-left (- (get debt-to-raise auction) (get total-debt-burned auction)))
```
Q/REQ: maybe we can change this for `(>= (get total-debt-burned auction) (get debt-to-raise auction))`. I don't know if it can ever happen, but if this would be a negative number, this throws an error and as a result, an auction can never be finalised

philipdesmedt reviewed (changes requested):
Did another review.. need to go over the liquidation pool and liquidation rewards contract again..

One general concern would be that the functions are costly. We have to do some cost analysis on it and check how many TXs of each would fit in a block

philipdesmedt commented:
@nieldeckx @Macxim 
I reviewed this and fixed a few smaller issues that were currently there. still need to do:

1. I was the only address staking USDA and could add the DIKO rewards 3 times after a while (since a lot of blocks had passed). I did that 3 times (3x 44,233 DIKO) and after the 3rd time, I could claim 44K DIKO once. Shouldn't I be able to claim  all the DIKO (so not 1x 44K. but 3x 44K DIKO)?
2. I think the "Add DIKO rewards to pool" button is confusing. We can leave it out and automate it through a script. It has to be ran only once every 720 blocks, so not often.
3. We need to change the token address of STX-A and STX-B to .xstx-token on mainnet

I have only tested vaults that were not stacking so far, still need to test stacking vaults and xBTC vaults.

philipdesmedt commented:
@nieldeckx about 1 - here are the payloads of the 3 transactions to add DIKO to the rewards pool

```
{
  "tx_id": "0xaf8b6768d9d90366884be5cef343c0a99c9071e4d7e974bf1c5b34b8e02ce667",
  "nonce": 14,
  "fee_rate": "73080",
  "sender_address": "ST3KCNDSWZSFZCC6BE4VA9AXWXC9KEB16FBTRK36T",
  "sponsored": false,
  "post_condition_mode": "allow",
  "post_conditions": [],
  "anchor_mode": "any",
  "is_unanchored": false,
  "block_hash": "0x52e6a998ecd3c7856e206bdf3f1a05bb17e9d8b515f3dabb63eeb8cf6527ca92",
  "parent_block_hash": "0x15a4f4a8debb0c32be57519c17e4a3c24c840b50425a5b6ff5b5dd81ab9e1308",
  "block_height": 2470,
  "burn_block_time": 1648825799,
  "burn_block_time_iso": "2022-04-01T15:09:59.000Z",
  "parent_burn_block_time": 1648825794,
  "parent_burn_block_time_iso": "2022-04-01T15:09:54.000Z",
  "canonical": true,
  "tx_index": 1,
  "tx_status": "success",
  "tx_result": {
    "hex": "0x070100000000000000000000000a4c8448b8",
    "repr": "(ok u44233410744)"
  },
  "microblock_hash": "",
  "microblock_sequence": 2147483647,
  "microblock_canonical": true,
  "event_count": 4,
  "events": [
    {
      "event_index": 0,
      "event_type": "smart_contract_log",
      "tx_id": "0xaf8b6768d9d90366884be5cef343c0a99c9071e4d7e974bf1c5b34b8e02ce667",
      "contract_log": {
        "contract_id": "ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-dao",
        "topic": "print",
        "value": {
          "hex": "0x0c0000000306616374696f6e0d000000066d696e74656404646174610c0000000206616d6f756e740100000000000000000000000a4c8448b809726563697069656e74061a6d78de7b0625dfbfc16c3a8a5735f6dc3dc3f2ce2661726b6164696b6f2d6c69717569646174696f6e2d726577617264732d64696b6f2d76312d3104747970650d00000005746f6b656e",
          "repr": "(tuple (action \"minted\") (data (tuple (amount u44233410744) (recipient ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-liquidation-rewards-diko-v1-1))) (type \"token\"))"
        }
      }
    },
    {
      "event_index": 1,
      "event_type": "fungible_token_asset",
      "tx_id": "0xaf8b6768d9d90366884be5cef343c0a99c9071e4d7e974bf1c5b34b8e02ce667",
      "asset": {
        "asset_event_type": "mint",
        "asset_id": "ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-token::diko",
        "sender": "",
        "recipient": "ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-liquidation-rewards-diko-v1-1",
        "amount": "44233410744"
      }
    },
    {
      "event_index": 2,
      "event_type": "fungible_token_asset",
      "tx_id": "0xaf8b6768d9d90366884be5cef343c0a99c9071e4d7e974bf1c5b34b8e02ce667",
      "asset": {
        "asset_event_type": "transfer",
        "asset_id": "ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-token::diko",
        "sender": "ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-liquidation-rewards-diko-v1-1",
        "recipient": "ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-liquidation-rewards-v1-1",
        "amount": "44233410744"
      }
    },
    {
      "event_index": 3,
      "event_type": "smart_contract_log",
      "tx_id": "0xaf8b6768d9d90366884be5cef343c0a99c9071e4d7e974bf1c5b34b8e02ce667",
      "contract_log": {
        "contract_id": "ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-token",
        "topic": "print",
        "value": {
          "hex": "0x09",
          "repr": "none"
        }
      }
    }
  ],
  "execution_cost_read_count": 53,
  "execution_cost_read_length": 78587,
  "execution_cost_runtime": 83680000,
  "execution_cost_write_count": 7,
  "execution_cost_write_length": 357,
  "tx_type": "contract_call",
  "contract_call": {
    "contract_id": "ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-liquidation-rewards-diko-v1-1",
    "function_name": "add-rewards",
    "function_signature": "(define-public (add-rewards (liquidation-rewards trait_reference)))",
    "function_args": [
      {
        "hex": "0x061a6d78de7b0625dfbfc16c3a8a5735f6dc3dc3f2ce2161726b6164696b6f2d6c69717569646174696f6e2d726577617264732d76312d31",
        "repr": "ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-liquidation-rewards-v1-1",
        "name": "liquidation-rewards",
        "type": "trait_reference"
      }
    ]
  }
}
```


second
```
{
  "tx_id": "0xa83f7c5a67504af448e043f5700f0bfb5d9d4439ae65ba3334937167d4c41eb0",
  "nonce": 13,
  "fee_rate": "99200",
  "sender_address": "ST3KCNDSWZSFZCC6BE4VA9AXWXC9KEB16FBTRK36T",
  "sponsored": false,
  "post_condition_mode": "allow",
  "post_conditions": [],
  "anchor_mode": "any",
  "is_unanchored": false,
  "block_hash": "0xc323a5137c74d6865e8ac5ad8a096619923d4b5861691b6429cd2a7e285b936a",
  "parent_block_hash": "0xb9e30945c9b3246efb12f6e660aff59e37104256c5c9a805d20592bd9efd365a",
  "block_height": 2467,
  "burn_block_time": 1648825783,
  "burn_block_time_iso": "2022-04-01T15:09:43.000Z",
  "parent_burn_block_time": 1648825778,
  "parent_burn_block_time_iso": "2022-04-01T15:09:38.000Z",
  "canonical": true,
  "tx_index": 1,
  "tx_status": "success",
  "tx_result": {
    "hex": "0x070100000000000000000000000a4c8448b8",
    "repr": "(ok u44233410744)"
  },
  "microblock_hash": "",
  "microblock_sequence": 2147483647,
  "microblock_canonical": true,
  "event_count": 4,
  "events": [
    {
      "event_index": 0,
      "event_type": "smart_contract_log",
      "tx_id": "0xa83f7c5a67504af448e043f5700f0bfb5d9d4439ae65ba3334937167d4c41eb0",
      "contract_log": {
        "contract_id": "ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-dao",
        "topic": "print",
        "value": {
          "hex": "0x0c0000000306616374696f6e0d000000066d696e74656404646174610c0000000206616d6f756e740100000000000000000000000a4c8448b809726563697069656e74061a6d78de7b0625dfbfc16c3a8a5735f6dc3dc3f2ce2661726b6164696b6f2d6c69717569646174696f6e2d726577617264732d64696b6f2d76312d3104747970650d00000005746f6b656e",
          "repr": "(tuple (action \"minted\") (data (tuple (amount u44233410744) (recipient ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-liquidation-rewards-diko-v1-1))) (type \"token\"))"
        }
      }
    },
    {
      "event_index": 1,
      "event_type": "fungible_token_asset",
      "tx_id": "0xa83f7c5a67504af448e043f5700f0bfb5d9d4439ae65ba3334937167d4c41eb0",
      "asset": {
        "asset_event_type": "mint",
        "asset_id": "ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-token::diko",
        "sender": "",
        "recipient": "ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-liquidation-rewards-diko-v1-1",
        "amount": "44233410744"
      }
    },
    {
      "event_index": 2,
      "event_type": "fungible_token_asset",
      "tx_id": "0xa83f7c5a67504af448e043f5700f0bfb5d9d4439ae65ba3334937167d4c41eb0",
      "asset": {
        "asset_event_type": "transfer",
        "asset_id": "ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-token::diko",
        "sender": "ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-liquidation-rewards-diko-v1-1",
        "recipient": "ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-liquidation-rewards-v1-1",
        "amount": "44233410744"
      }
    },
    {
      "event_index": 3,
      "event_type": "smart_contract_log",
      "tx_id": "0xa83f7c5a67504af448e043f5700f0bfb5d9d4439ae65ba3334937167d4c41eb0",
      "contract_log": {
        "contract_id": "ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-token",
        "topic": "print",
        "value": {
          "hex": "0x09",
          "repr": "none"
        }
      }
    }
  ],
  "execution_cost_read_count": 53,
  "execution_cost_read_length": 78587,
  "execution_cost_runtime": 83680000,
  "execution_cost_write_count": 7,
  "execution_cost_write_length": 357,
  "tx_type": "contract_call",
  "contract_call": {
    "contract_id": "ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-liquidation-rewards-diko-v1-1",
    "function_name": "add-rewards",
    "function_signature": "(define-public (add-rewards (liquidation-rewards trait_reference)))",
    "function_args": [
      {
        "hex": "0x061a6d78de7b0625dfbfc16c3a8a5735f6dc3dc3f2ce2161726b6164696b6f2d6c69717569646174696f6e2d726577617264732d76312d31",
        "repr": "ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-liquidation-rewards-v1-1",
        "name": "liquidation-rewards",
        "type": "trait_reference"
      }
    ]
  }
}
```


third
```
{
  "tx_id": "0x70d10e3528d3f7bd62ed60cecb90ec82149bb41f945d39c89b8a2031d715abea",
  "nonce": 12,
  "fee_rate": "99200",
  "sender_address": "ST3KCNDSWZSFZCC6BE4VA9AXWXC9KEB16FBTRK36T",
  "sponsored": false,
  "post_condition_mode": "allow",
  "post_conditions": [],
  "anchor_mode": "any",
  "is_unanchored": false,
  "block_hash": "0xc03be49a05731379d066b91d4169838c685e91709c6f361e9071740b5d3128dc",
  "parent_block_hash": "0x1bdccaa0e14a6713a085a7ca0b29d6639891d884f3a00a1526fa85972fc6f9b6",
  "block_height": 2465,
  "burn_block_time": 1648825773,
  "burn_block_time_iso": "2022-04-01T15:09:33.000Z",
  "parent_burn_block_time": 1648825768,
  "parent_burn_block_time_iso": "2022-04-01T15:09:28.000Z",
  "canonical": true,
  "tx_index": 1,
  "tx_status": "success",
  "tx_result": {
    "hex": "0x070100000000000000000000000a4c8448b8",
    "repr": "(ok u44233410744)"
  },
  "microblock_hash": "",
  "microblock_sequence": 2147483647,
  "microblock_canonical": true,
  "event_count": 4,
  "events": [
    {
      "event_index": 0,
      "event_type": "smart_contract_log",
      "tx_id": "0x70d10e3528d3f7bd62ed60cecb90ec82149bb41f945d39c89b8a2031d715abea",
      "contract_log": {
        "contract_id": "ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-dao",
        "topic": "print",
        "value": {
          "hex": "0x0c0000000306616374696f6e0d000000066d696e74656404646174610c0000000206616d6f756e740100000000000000000000000a4c8448b809726563697069656e74061a6d78de7b0625dfbfc16c3a8a5735f6dc3dc3f2ce2661726b6164696b6f2d6c69717569646174696f6e2d726577617264732d64696b6f2d76312d3104747970650d00000005746f6b656e",
          "repr": "(tuple (action \"minted\") (data (tuple (amount u44233410744) (recipient ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-liquidation-rewards-diko-v1-1))) (type \"token\"))"
        }
      }
    },
    {
      "event_index": 1,
      "event_type": "fungible_token_asset",
      "tx_id": "0x70d10e3528d3f7bd62ed60cecb90ec82149bb41f945d39c89b8a2031d715abea",
      "asset": {
        "asset_event_type": "mint",
        "asset_id": "ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-token::diko",
        "sender": "",
        "recipient": "ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-liquidation-rewards-diko-v1-1",
        "amount": "44233410744"
      }
    },
    {
      "event_index": 2,
      "event_type": "fungible_token_asset",
      "tx_id": "0x70d10e3528d3f7bd62ed60cecb90ec82149bb41f945d39c89b8a2031d715abea",
      "asset": {
        "asset_event_type": "transfer",
        "asset_id": "ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-token::diko",
        "sender": "ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-liquidation-rewards-diko-v1-1",
        "recipient": "ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-liquidation-rewards-v1-1",
        "amount": "44233410744"
      }
    },
    {
      "event_index": 3,
      "event_type": "smart_contract_log",
      "tx_id": "0x70d10e3528d3f7bd62ed60cecb90ec82149bb41f945d39c89b8a2031d715abea",
      "contract_log": {
        "contract_id": "ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-token",
        "topic": "print",
        "value": {
          "hex": "0x09",
          "repr": "none"
        }
      }
    }
  ],
  "execution_cost_read_count": 53,
  "execution_cost_read_length": 78587,
  "execution_cost_runtime": 83680000,
  "execution_cost_write_count": 7,
  "execution_cost_write_length": 357,
  "tx_type": "contract_call",
  "contract_call": {
    "contract_id": "ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-liquidation-rewards-diko-v1-1",
    "function_name": "add-rewards",
    "function_signature": "(define-public (add-rewards (liquidation-rewards trait_reference)))",
    "function_args": [
      {
        "hex": "0x061a6d78de7b0625dfbfc16c3a8a5735f6dc3dc3f2ce2161726b6164696b6f2d6c69717569646174696f6e2d726577617264732d76312d31",
        "repr": "ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM.arkadiko-liquidation-rewards-v1-1",
        "name": "liquidation-rewards",
        "type": "trait_reference"
      }
    ]
  }
}
```

nieldeckx commented:
> @nieldeckx @Macxim I reviewed this and fixed a few smaller issues that were currently there. still need to do:
> 
> 1. I was the only address staking USDA and could add the DIKO rewards 3 times after a while (since a lot of blocks had passed). I did that 3 times (3x 44,233 DIKO) and after the 3rd time, I could claim 44K DIKO once. Shouldn't I be able to claim  all the DIKO (so not 1x 44K. but 3x 44K DIKO)?
> 2. I think the "Add DIKO rewards to pool" button is confusing. We can leave it out and automate it through a script (just removing the button and keep in the UI how much DIKO will be emitted). It has to be ran only once every 720 blocks, so not often.
> 3. We need to change the token address of STX-A and STX-B to .xstx-token on mainnet
> 
> I have only tested vaults that were not stacking so far, still need to test stacking vaults and xBTC vaults.

1. I am not able to reproduce this. Tests show it's working. If epoch is not over and you try to add DIKO rewards you should get an error (ERR-EPOCH-NOT-ENDED u34002). The reason you could not claim any rewards might be because you have deposited USDA in the liquidation pool after the second epoch?

2. Agree. Removed the button and added script to add DIKO rewards (add-liquidation-pool-rewards.js)

philipdesmedt commented:
@nieldeckx yeah - perhaps I deposited USDA after - does it reset if you deposit twice? I remember adding USDA twice... but since I was always the only depositor, shouldn't I be able to claim all DIKO regardless? 🤔

nieldeckx commented:
> @nieldeckx yeah - perhaps I deposited USDA after - does it reset if you deposit twice? I remember adding USDA twice... but since I was always the only depositor, shouldn't I be able to claim all DIKO regardless? 🤔

It should not reset. When adding diko rewards at the end of the epoch, the system will take into account your USDA deposits at the start of the epoch.

In production, the epoch length is 5 days and we can start the first epoch 10 days after deployment, so there are 5 days in which people can add USDA. This way there will always be some USDA when distributing DIKO rewards, and the scenario you are describing should not happen. That's the idea at least :)

philipdesmedt commented on clarity/contracts/arkadiko-auction-engine-v4-1.clar:
```diff
@@ -0,0 +1,440 @@
+    (asserts! (is-eq (contract-of coll-type) (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "collateral-types"))) (err ERR-NOT-AUTHORIZED))
+
+    ;; Sell off DIKO if needed
+    (try! (sell-diko auction-id))
```
REQ: I think there's something odd going on here.

Imagine the following scenario:
1. A vault gets liquidated but there is not enough USDA to buy all collateral in the fund
2. Someone calls this immediately after. Since there is not enough USDA, it mints DIKO and sells that off

This is a bit odd since we should only mint and sell DIKO if the total collateral amount in the vault is not sufficient to pay back the debt, and not whenever there is not sufficient USDA. In other words, DIKO minting should be considered in isolation, per vault. Imagine a vault of 1M STX needs to get liquidated in current model, this would completely destroy the DIKO price.

nieldeckx replied:
On L321 in function `sell-diko` we check if all collateral is sold. Only if that's the case, DIKO will be sold.

philipdesmedt replied:
Oh nice, I missed that. Should we do something like >= (or <=) instead of `is-eq`? Not sure if something can go wrong but it seems a bit more loose and less error prone

philipdesmedt commented on clarity/contracts/arkadiko-auction-engine-v4-1.clar:
```diff
@@ -0,0 +1,440 @@
+    (burn-amounts (unwrap-panic (burn-usda-amount auction-id oracle liquidation-pool)))
+    (usda-to-use (get usda-to-use burn-amounts))
+    (collateral-sold (get collateral-to-sell burn-amounts))
+    (collateral-fee (/ (* collateral-sold (var-get auction-fee)) u10000))
```
Q: how do we get access to this collateral?

nieldeckx replied:
Fees will be collected in the contract. Using the 'withdraw-fees' method we can withdraw thes.

philipdesmedt commented on clarity/contracts/arkadiko-auction-engine-v4-1.clar:
```diff
@@ -0,0 +1,440 @@
+;; @param auction-id; the ID of the auction for which the collateral amount should be calculated
+;; @param oracle; the oracle implementation that provides the on-chain price
+;; @param liquidation-pool; the pool that holds USDA to use in liquidations
+(define-public (burn-usda-amount
```
Q: this cannot be read-only I assume due to the traits?

nieldeckx replied:
Correct.

philipdesmedt commented on clarity/contracts/arkadiko-auction-engine-v4-1.clar:
```diff
@@ -0,0 +1,440 @@
+    (auction (get-auction-by-id auction-id))
+    (debt-left (- (get debt-to-raise auction) (get total-debt-burned auction)))
+    (collateral-left (- (get collateral-amount auction) (get total-collateral-sold auction)))
+    (collateral-price (unwrap-panic (get-collateral-discounted-price oracle auction-id)))
```
NTH: it would be great if we can get rid of the `unwrap-panic` here and just use `unwrap!`

nieldeckx replied:
Updated.

philipdesmedt commented on clarity/contracts/arkadiko-liquidation-rewards-diko-v1-1.clar:
```diff
@@ -0,0 +1,101 @@
+(define-read-only (get-rewards-to-add)
+  (let (
+    (total-block-rewards (contract-call? .arkadiko-diko-guardian-v1-1 get-staking-rewards-per-block))
+    (rewards-to-add (/ (* total-block-rewards (var-get blocks-per-epoch) (var-get epoch-rate)) u1000000))
```
Q: is this actually correct? Doesn't the reward go down slightly each block? Or should we ignore that as a rounding number?

nieldeckx replied:
It is slightly off indeed. To calculate the exact rewards we would need to loop over every block which is not feasible. The approximation should be fine if the periods are not too long. We have the same approach for staking rewards.

philipdesmedt commented on clarity/contracts/arkadiko-auction-engine-v4-1.clar:
```diff
@@ -378,7 +378,7 @@
     (debt-left (- (get debt-to-raise auction) (get total-debt-burned auction)))
     (collateral-left (- (get collateral-amount auction) (get total-collateral-sold auction)))
-    (collateral-price (unwrap-panic (get-collateral-discounted-price oracle auction-id)))
+    (collateral-price (unwrap! (get-collateral-discounted-price oracle auction-id)))
```
@nieldeckx you need to add a specific error here at the end

nieldeckx replied:
Oops, done. :)
